miden/examples/fri/fri.masm

```diff
@@ -0,0 +1,328 @@
+        u32checked_and          #[x0, x>>1,r,g,...]   where x0 := x&1
+        movup.2  dup            #[r,r,x0, x>>1 ,g,...]
+        dup.4  mul              #[gr,r,x0, x>>1 ,g,...]
+        movup.2                 #[x0,gr,r, x>>1 ,g,...]
+        cdrop                   #[g^x0, x>>1 ,g,...]  
+        movup.2 dup mul         #[g^2,g^x0, x>>1 ,...]
+        swap                    #[g^x0,g^2, x>>1 ,...]
+    end
+    swap drop swap drop
+end
+
+# Input: [t_d,e1,e0,p,d,poe,..]
+# Output: [d,p,C,t_d,e1,e0,poe,a1,a0,..]
+proc.prepare_next
+    adv_push.2                                #[a1,a0,t_d,e1,e0,p,d,poe,..]
+    swapw                                     #[e0,p,d,poe,a,a,t_d,e1,..]
+    push.0.0.0.0 adv_loadw                    #[C,e0,p,d,poe,a,a,t_d,e1,..]
+    movup.5 movup.6                           #[d,p,C,e0,poe,a,a,t_d,e1,..]
+    movup.11                                  #[e1,d,p,C,e0,poe,a,a,t_d,..]
+    movdn.6                                   #[d,p,C,e1,e0,poe,a,a,t_d,..]
+    movup.11                                  #[t_d,d,p,C,e1,e0,poe,a,a,..]
+    movdn.6                                   #[d,p,C,t_d,e1,e0,poe,a1,a0,..]
+end
+
+# Prepare the stack by injecting advice inputs from the advice tape. Should check the relevant
+# conditions on the loaded inputs.
+# Input(Advice tape): [e1,e0,a1,a0,t_d,C,p,d,g,...]
+# Output: [d,p,C,t_d,e1,e0,poe,a1,a0,...]
+proc.prepare_initial
+    push.0.0.0.0 adv_loadw
+    adv_push.1
+    push.0.0.0.0 adv_loadw
+    adv_push.2              #[d,p,C,...]
+    dup.1                   #[p,d,p,C,...]
+    # Here we should probably check that p < d in u32 and that t_d is in u32 and that 2^t_d = d 
+    adv_push.1              #[g,p,d,p,C,...]
+    exec.exp_g_x            #[poe,d,p,C,t_d,e1,e0,a1,a0,...]
+    movdn.9                 #[d,p,C,t_d,e1,e0,poe,a1,a0,...]
+end
+
+# Prepare the stack for the final remainder check to be executed once all queries
+# have been processed. Assumes that the remainder has degree 0. In the case of 
+# degree 2 we will need p and poe on the stack.
+# Input: [t_d,e1,e0,p,d,poe,..]
+# Output: #[e1,e0,..]
+proc.prepare_final
+    repeat.3
+        movup.5
+    end
+    dropw
+end
+
+# Input: [e1,e0,e1,e0,e1,e0,e1,e0,e1,e0,e1,e0,...]
+# Output: []
+proc.verify_remainder
+    repeat.4
+        movup.2                     #[e1,v1,v0,e0,v1,v0,e1,e0,...]
+        assert_eq                   #[v0,e0,v1,v0,e1,e0,...]
+        assert_eq                   #[v1,v0,e1,e0,...]
+    end
+end
+
+
+# Executes a full FRI verification for a single query.
+proc.verify_query_full
+    exec.prepare_initial    #[d,p,C,t_d,e1,e0,poe,a1,a0,...]
+    dup push.16 neq         #[?,d,p,C,t_d,e1,e0,poe,a1,a0,...]
+
+    while.true
+        exec.verify_query_layer
+        exec.prepare_next
+        dup push.16 neq
+    end
+
+    exec.verify_query_layer
+    exec.prepare_final
+
+    # Final configuration
+    # [e1,e0,..]
+end
+
+# Verify a FRI proof
+proc.verify_fri
+    repeat.32
```

Comment:
Does 32 specify the number of queries?

Reply:
Yes, it does but thinking about your second point above this construction makes little sense now.
To your second point. I have indeed missed the consistency between the layer commitments from one query to the next and I agree that the most straightforward way to solving this is by using memory reads and writes. The question then would be: how to lay out the advice input so as to optimize the reads/writes and minimize some of the repeated computations that are performed for each query.
As a start, I am thinking of the following:
- To memory at the beginning: Commitments, domain generator, domain size, log(domain size), alphas
- To stack from advice tape gradually: Query values
Does this match, approximately, what you were thinking?

Reply:
Yes - that's how I was thinking about it as well. A couple more thoughts:

1. We can probably pre-compute domain generators and other such common variables for all layers right from the start (and put them in sequential memory locations). This way, with every layer we can just increment the memory pointer by one and get the relevant values.
2. For query values, we can probably use the new key-value map mechanism via the `adv.keval` instruction.